lib/utils/string_extension.dart

```diff
@@ -89,4 +89,34 @@ extension StringCasingExtension on String {
     final String? firstValidLink = matches.firstWhere((link) => AnyLinkPreview.isValidLink(link!));
     return firstValidLink;
   }
+
+  // Removes markdowned links from a string based on the unformatted text
+  // Workaround for content['formatted_body'] which formats urls in a way that makes them unusable
+  String unMarkdownLinks(String unformattedText) {
+    final RegExp regex = RegExp(r'https:\/\/[^\s]+');
+
+    final Iterable<Match> formattedLinksMatches = regex.allMatches(this);
+    final Iterable<Match> unformattedLinksMatches = regex.allMatches(unformattedText);
+
+    if (formattedLinksMatches.isEmpty ||
+        unformattedLinksMatches.isEmpty ||
+        formattedLinksMatches.length != unformattedLinksMatches.length) {
+      return this;
+    }
+
+    var unMarkdownedText = this;
+
+    // Replace respectively all formatted links with unformatted links
+    for (int i = 0; i < formattedLinksMatches.length; i++) {
+      final Match formattedLinkMatch = formattedLinksMatches.elementAt(i);
+      final Match unformattedLinkMatch = unformattedLinksMatches.elementAt(i);
+
+      final String formattedLink = formattedLinkMatch.group(0)!;
+      final String unformattedLink = unformattedLinkMatch.group(0)!;
+
+      unMarkdownedText = unMarkdownedText.replaceFirst(formattedLink, unformattedLink);
+    }
+
+    return unMarkdownedText;
+  }
```

Comment:
1. Replace `this` keyword with the `unformattedText` parameter, as it should be used instead.
2. Change the for loop to a while loop using iterators to ensure that we iterate over both sets of matches simultaneously.
3. Change `elementAt(i)` to `current`
3. Replacet `this` with `unformattedText`  in the line where you replace the formatted link with the unformatted link.

Reply:
1. I want to keep formatted text / markdown on everything except URLs. IMO it's easier to start from formatted and replace only links so we keep everything else

I'll work on 2, 3

Reply:
```
while (formattedIterator.moveNext() && unformattedIterator.moveNext()) {
    final Match formattedLinkMatch = formattedIterator.current;
    final Match unformattedLinkMatch = unformattedIterator.current;

    final String formattedLink = formattedLinkMatch.group(0)!;
    final String unformattedLink = unformattedLinkMatch.group(0)!;

    unMarkdownedText = unMarkdownedText.replaceFirst(formattedLink, unformattedLink);
  }
```

WDYT

Reply:
how about:
String unMarkdownLinks(String unformattedText) {
  final RegExp regex = RegExp(r'https:\/\/[^\s]+');

  final Iterable<Match> formattedLinksMatches = regex.allMatches(unformattedText);
  final List<String> formattedLinks = formattedLinksMatches.map((match) => match.group(0)!).toList();

  if (formattedLinks.isEmpty) {
    return unformattedText;
  }

  String unMarkdownedText = unformattedText;

  for (String formattedLink in formattedLinks) {
    final String unformattedLink = formattedLink;

    unMarkdownedText = unMarkdownedText.replaceFirst(formattedLink, unformattedLink);
  }

  return unMarkdownedText;
}

Reply:
I'll put @nqhhdev s version, it's cleaner and safer